processing/src/main/java/io/druid/segment/IndexMerger.java

```diff
@@ -271,19 +273,52 @@ public File merge(
     );
   }
 
+  private List<String> getLongestSharedDimOrder(List<IndexableAdapter> indexes)
+  {
+    int maxSize = 0;
+    Iterable<String> orderingCandidate = null;
+    for (IndexableAdapter index : indexes) {
+      int iterSize = Iterators.size(index.getDimensionNames().iterator());
+      if (iterSize > maxSize) {
+        maxSize = iterSize;
+        orderingCandidate = index.getDimensionNames();
+      }
+    }
+
+    if (orderingCandidate == null) {
+      return null;
+    }
+
+    for (IndexableAdapter index : indexes) {
+      Iterator<String> candidateIter = orderingCandidate.iterator();
+      for (String matchDim : index.getDimensionNames()) {
+        boolean matched = false;
+        while (candidateIter.hasNext()) {
```

Comment:
@himanshug I don't think a Set would be helpful here, since the ordering of dimensions matters, it's not enough to check for presence in the candidate.

It could be useful though to build a map of dimension->list index for the dimensions in the candidate, and the per-index check could then iterate through the dimensions present in each index, checking that those dimensions exist in the candidate and their list indices always increase through the iterations

It would reduce the number of dimension list iterations when some dimension lists are significantly smaller than the candidate list, but not so much when the lists are similarly sized.